### PR TITLE
Audio system overhaul

### PR DESCRIPTION
**Pull request by drashna**
This builds upon #6165 and #6123. 

## Description

Most of this work was done by @JohSchneider and @jackhumbert, and credit goes to them. As well as @spidey3 for finding and fixing the AVR audio bug here. 

* Refactors the audio into AVR and ARM specific implementations
* Moves common code into a single file for manageability, 
* Allows for configurable DAC support (eg, you can use one or both pins), 
* PWM support on ChibiOS for chips that don't have DAC support
* Implement/refactor vibrato
* Implement/refactor timbre
* Implement rest/pause between two same notes of a melody

Additionally. this does include the ZSA fixes for basic dac support (disabling pins when off). 

Glissando refactor needed, but beyond my experience to do (from 6165's to-do)

## Types of Changes

- [x] Core
- [x] Bugfix
- [x] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [x] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: quantum/audio/musical_notes.h [hunk 1]
```diff
@@ -27,19 +26,23 @@
...
 #define MUSICAL_NOTE(note, duration) \
     { (NOTE##note), duration }
+
 #define BREVE_NOTE(note) MUSICAL_NOTE(note, 128)
 #define WHOLE_NOTE(note) MUSICAL_NOTE(note, 64)
 #define HALF_NOTE(note) MUSICAL_NOTE(note, 32)
 #define QUARTER_NOTE(note) MUSICAL_NOTE(note, 16)
 #define EIGHTH_NOTE(note) MUSICAL_NOTE(note, 8)
 #define SIXTEENTH_NOTE(note) MUSICAL_NOTE(note, 4)
+#define THIRTYTWOTH_NOTE(note) MUSICAL_NOTE(note, 2)
```
Comment (jackhumbert) [inline quantum/audio/musical_notes.h:36]:
```suggestion
#define THIRTYSECOND_NOTE(note) MUSICAL_NOTE(note, 2)
```

File: quantum/audio/musical_notes.h [hunk 2]
```diff
@@ -27,19 +26,23 @@
...
 
 #define BREVE_DOT_NOTE(note) MUSICAL_NOTE(note, 128 + 64)
 #define WHOLE_DOT_NOTE(note) MUSICAL_NOTE(note, 64 + 32)
 #define HALF_DOT_NOTE(note) MUSICAL_NOTE(note, 32 + 16)
 #define QUARTER_DOT_NOTE(note) MUSICAL_NOTE(note, 16 + 8)
 #define EIGHTH_DOT_NOTE(note) MUSICAL_NOTE(note, 8 + 4)
 #define SIXTEENTH_DOT_NOTE(note) MUSICAL_NOTE(note, 4 + 2)
```
Comment (jackhumbert) [inline quantum/audio/musical_notes.h:43]:
```suggestion
#define SIXTEENTH_DOT_NOTE(note) MUSICAL_NOTE(note, 4 + 2)
#define THIRTYSECOND_DOT_NOTE(note) MUSICAL_NOTE(note, 2 + 1)
```

File: quantum/audio/musical_notes.h [hunk 3]
```diff
@@ -58,47 +62,41 @@
```
Comment (jackhumbert) [inline quantum/audio/musical_notes.h:62]:
```suggestion
#define TD_NOTE(n) THIRTYSECOND_DOT_NOTE(n)
```

File: quantum/audio/musical_notes.h [hunk 4]
```diff
@@ -49,6 +52,7 @@
 #define Q__NOTE(n) QUARTER_NOTE(n)
 #define E__NOTE(n) EIGHTH_NOTE(n)
 #define S__NOTE(n) SIXTEENTH_NOTE(n)
+#define T__NOTE(n) THIRTYTWOTH_NOTE(n)
```
Comment (jackhumbert) [inline quantum/audio/musical_notes.h:55]:
```suggestion
#define T__NOTE(n) THIRTYSECOND_NOTE(n)
```

File: util/audio_generate_dac_lut.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
+#!/usr/bin/env python3
```
Comment (jackhumbert) [inline util/audio_generate_dac_lut.py:1]: I think this needs a copyright header, right?

Comment (drashna) [reply util/audio_generate_dac_lut.py:1]: @JohSchneider Just wanted to make sure, was this written by you?

**Review by jackhumbert — APPROVED** [review]

File: docs/config_options.md [hunk 6]
```diff
@@ -67,16 +67,22 @@ This is a C header file that is one of the first things included, and will persi
...
+  * enables audio on pin B5 (duophony is enables if one of B pins is enabled along with one of C pins)
+  * Depricated. Use `#define AUDIO_PIN B5`, or use `#define AUDIO_PIN_ALT B5` if a `C` pin is enabled with `AUDIO_PIN`
 * `#define B6_AUDIO`
-  * enables audio on pin B6 (duophony is enables if one of B[5-7]\_AUDIO is enabled along with one of C[4-6]\_AUDIO)
+  * enables audio on pin B5 (duophony is enables if one of B pins is enabled along with one of C pins)
+  * Depricated. Use `#define AUDIO_PIN B6`, or use `#define AUDIO_PIN_ALT B6` if a `C` pin is enabled with `AUDIO_PIN`
 * `#define B7_AUDIO`
-  * enables audio on pin B7 (duophony is enables if one of B[5-7]\_AUDIO is enabled along with one of C[4-6]\_AUDIO)
+  * enables audio on pin B5 (duophony is enables if one of B pins is enabled along with one of C pins)
+  * Depricated. Use `#define AUDIO_PIN B7`, or use `#define AUDIO_PIN_ALT B7` if a `C` pin is enabled with `AUDIO_PIN`
```
Comment (spidey3) [inline docs/config_options.md:85]:
```suggestion
  * Deprecated. Use `#define AUDIO_PIN C4`
* `#define C5_AUDIO`	* `#define C5_AUDIO`
  * enables audio on pin C5	  * enables audio on pin C5
  * Deprecated. Use `#define AUDIO_PIN C5`
* `#define C6_AUDIO`	* `#define C6_AUDIO`
  * enables audio on pin C6	  * enables audio on pin C6
  * Deprecated. Use `#define AUDIO_PIN C6`
* `#define B5_AUDIO`	* `#define B5_AUDIO`
  * enables audio on pin B5 (duophony is enabled if one of B pins is enabled along with one of C pins)
  * Deprecated. Use `#define AUDIO_PIN B5`, or use `#define AUDIO_PIN_ALT B5` if a `C` pin is enabled with `AUDIO_PIN`
* `#define B6_AUDIO`	* `#define B6_AUDIO`
  * enables audio on pin B6 (duophony is enabled if one of B[5-7]\_AUDIO is enabled along with one of C[4-6]\_AUDIO)	  * enables audio on pin B5 (duophony is enabled if one of B pins is enabled along with one of C pins)
  * Deprecated. Use `#define AUDIO_PIN B6`, or use `#define AUDIO_PIN_ALT B6` if a `C` pin is enabled with `AUDIO_PIN`
* `#define B7_AUDIO`	* `#define B7_AUDIO`
  * enables audio on pin B7 (duophony is enabled if one of B[5-7]\_AUDIO is enabled along with one of C[4-6]\_AUDIO)	  * enables audio on pin B5 (duophony is enabled if one of B pins is enabled along with one of C pins)
  * Deprecated. Use `#define AUDIO_PIN B7`, or use `#define AUDIO_PIN_ALT B7` if a `C` pin is enabled with `AUDIO_PIN`
```

Comment (spidey3) [inline docs/config_options.md:85]:
```suggestion
  * Deprecated. Use `#define AUDIO_PIN C4`
* `#define C5_AUDIO`	* `#define C5_AUDIO`
  * enables audio on pin C5	  * enables audio on pin C5
  * Deprecated. Use `#define AUDIO_PIN C5`
* `#define C6_AUDIO`	* `#define C6_AUDIO`
  * enables audio on pin C6	  * enables audio on pin C6
  * Deprecated. Use `#define AUDIO_PIN C6`
* `#define B5_AUDIO`	* `#define B5_AUDIO`
  * enables audio on pin B5 (duophony is enabled if one of B pins is enabled along with one of C pins)
  * Deprecated. Use `#define AUDIO_PIN B5`, or use `#define AUDIO_PIN_ALT B5` if a `C` pin is enabled with `AUDIO_PIN`
* `#define B6_AUDIO`	* `#define B6_AUDIO`
  * enables audio on pin B6 (duophony is enabled if one of B[5-7]\_AUDIO is enabled along with one of C[4-6]\_AUDIO)	  * enables audio on pin B5 (duophony is enabled if one of B pins is enabled along with one of C pins)
  * Deprecated. Use `#define AUDIO_PIN B6`, or use `#define AUDIO_PIN_ALT B6` if a `C` pin is enabled with `AUDIO_PIN`
* `#define B7_AUDIO`	* `#define B7_AUDIO`
  * enables audio on pin B7 (duophony is enabled if one of B[5-7]\_AUDIO is enabled along with one of C[4-6]\_AUDIO)	  * enables audio on pin B5 (duophony is enabled if one of B pins is enabled along with one of C pins)
  * Deprecated. Use `#define AUDIO_PIN B7`, or use `#define AUDIO_PIN_ALT B7` if a `C` pin is enabled with `AUDIO_PIN`
```

Comment (drashna) [reply docs/config_options.md:85]: Fixed.

File: util/audio_generate_dac_lut.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+#
+# Copyright 2019 Jack Humbert
```
Comment (jackhumbert) [inline util/audio_generate_dac_lut.py:3]:
```suggestion
```

I know I didn't create/modify this

File: docs/audio_driver.md [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+//halconf.h:
+#define HAL_USE_DAC                 TRUE
+#define HAL_USE_GPT                 TRUE
+```
+
+``` c
+// mcuconf.h:
+#define STM32_DAC_USE_DAC1_CH1              TRUE
+#define STM32_DAC_USE_DAC1_CH2              TRUE
+#define STM32_GPT_USE_TIM6                  TRUE
```
Comment (tzarc) [inline docs/audio_driver.md:99]:
```suggestion
#include_next <mcuconf.h>
#undef STM32_DAC_USE_DAC1_CH1
#define STM32_DAC_USE_DAC1_CH1              TRUE
#undef STM32_DAC_USE_DAC1_CH2
#define STM32_DAC_USE_DAC1_CH2              TRUE
#undef STM32_GPT_USE_TIM6
#define STM32_GPT_USE_TIM6                  TRUE
```

File: docs/audio_driver.md [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+
+This driver uses the ChibiOS-PWM system to produce a square-wave on specific output pins that are connected to the PWM hardware.
+The hardware directly toggles the pin via its alternate function. See your MCU's data-sheet for which pin can be driven by what timer - looking for TIMx_CHy and the corresponding alternate function.
+
+A configuration example for the STM32F103C8 would be:
+``` c
+//halconf.h:
+#define HAL_USE_PWM                 TRUE
+#define HAL_USE_PAL                 TRUE
+#define HAL_USE_GPT                 TRUE
```
Comment (tzarc) [inline docs/audio_driver.md:148]:
```suggestion
#define HAL_USE_PWM                 TRUE
#define HAL_USE_PAL                 TRUE
#define HAL_USE_GPT                 TRUE
#include_next <halconf.h>
```

File: docs/audio_driver.md [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+```
+
+``` c
+// mcuconf.h:
+#define STM32_PWM_USE_TIM1                  TRUE
+#define STM32_GPT_USE_TIM4                  TRUE
```
Comment (tzarc) [inline docs/audio_driver.md:154]:
```suggestion
#include_next <mcuconf.h>
#undef STM32_PWM_USE_TIM1
#define STM32_PWM_USE_TIM1                  TRUE
#undef STM32_GPT_USE_TIM4
#define STM32_GPT_USE_TIM4                  TRUE
```

**Review by tzarc — CHANGES_REQUESTED** [review]

**Review by tzarc — APPROVED** [review]

**Pull request merged**